### ISSUE TITLE
Get all specific files from all Backups

### ISSUE THREAD

**Issue by Seoka** [borgbackup/borg#4841]
## Have you checked borgbackup docs, FAQ, and open Github issues?

Yes

## Is this a BUG / ISSUE report or a QUESTION?

Question/Feature Request

## System information. For client/server mode post info for both machines.

Can be uploaded after the Command is done.

## Describe the problem you're observing.

For Reasons I wanted to give someone all the backups I have of his files. After searching for about a day and not finding a valid, short resolution for me, I made a little horrible Bash Skript for this:
```bash
borg list --short | parallel --bar -P 1 --delay 2 --keep-order 'mkdir {}; cd {}; borg list --short ::{} | grep -E "{Name1}|{Name2}" | xargs -I++ borg extract --strip-components 4 ::{} ++'
```
This is some kind of awkward as this usecase seems to me, to be a frequently needed one.

I hope my bash command at least helps others in similar scenarios.

Regards
Seoka

**Comment by ThomasWaldmann** [comment]
maybe `borg mount <repo>` and then working on the dirs/files is easier?

guess the `parallel` in your command does not really lead to parallelism as borg locks the repo while working.

**Comment by Seoka** [comment]
yea, the `parallel` Command would work with multiple Threads, if I had not set the `-P 1` switch for turning Multi-threading off ("P" for Processes).
The `borg mount <repo>` version might even be faster (It was running about 6 hours on about 43 repositories). I did not think of that. At first I wanted to use the `borg export-tar` Command.

**Comment by ThomasWaldmann** [comment]
`borg mount -o versions` might be even better when interested in historical versions of same file(s).

You can limit the scope of borg mount using paths, --last, --first and other options, then it'll be a bit faster and less resource consuming.

**Comment by Seoka** [comment]
Perfekt! This solves my Problem.
(Suggestion: Change the Renaming of the actual files to one of these: 
* "{name_without_postfix}.{unixtimstamp}.{postfix}" 
* "{timestamp}.{postfix}"
* "{timestamp}_{full_file_name}"

this way you are not so likely to loose the information of the time. Timestamp could be anything from unix-timestamp to iso formated time.)
For all struggling with the command like me, my example: `borg mount -o versions :: /mnt/borg srv/docker/nextcloud/data/seoka/files/`

**Comment by ThomasWaldmann** [comment]
I don't remember precisely, but I guess:

Timestamp: the timestamp of the files is from the first backup archive when a change has been seen by borg. So, you have it there and the timestamp does not need to be duplicated into the generated filename.

Filename: the filename is the containing directory name, so no need for duplicating that either.

Archive name: that would be more interesting indeed, but a specific content can be in multiple archive, so (like with the timestamp), we could only use the **first** archivename where a new file content has been seen. I also imagined a mode where the directory contents are not just normal numbered versions, but symlinks that link into a per-archive directory structure, like:

- directory text.pdf
  - symlink 1 --> ../../archive1/.../text.pdf
  - symlink 2 --> ../../archive42/.../text.pdf

So that would be kind of a mixture between a normal `borg mount`'s directory structure and a `borg mount -o versions`'s structure.
Just an idea, not implemented yet.

**Comment by Seoka** [comment]
As this is of topic, I think this discussion should not be within this section. 
But since I am already replying, my thoughts:
You are right on the Timestamp and Filename duplication thoughts. Numbered symlinks might be, depending on the kind of implementation, pretty power consuming as you would need to mount a lot? 
```bash
root@seoka ~ $ borg list | wc -l && borg info --last 1 | tail -7
56
------------------------------------------------------------------------------
                       Original size      Compressed size    Deduplicated size
This archive:              178.11 GB            166.71 GB             35.15 MB
All archives:                9.89 TB              9.22 TB            211.60 GB

                       Unique chunks         Total chunks
Chunk index:                  203778              7173777

```
Thinking of mounting - lets say 50% - 5TB on my system makes me doubt it can handle it. -- unless it is hot loaded on access of course --

I actually like your idea of using the first archive name where a new file content has been seen. This way it is possible to load the whole environment from a specific file version.

**Comment by ThomasWaldmann** [comment]
Not sure I understand what problem you are referring to.

If the feature I described was implemented, of course it would need to create a lot of symlinks - but with the current `versions` implementation, the number of fs objects (numbered files) is the same amount (symlink source). The fs objects pointed to by the symlinks would add to that. But you also have that if you mount a whole repo and list all toplevel directories - it will also load all archives then.
